### PR TITLE
1412 local normalized cross correlation

### PR DESCRIPTION
**Pull request by kate-sann5100**
Fixes # 1412.

### Description
A few sentences describing the changes proposed in this pull request.

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: tests/test_local_normalized_cross_correlation_loss.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        {
+            "input": torch.arange(0, 3, dtype=torch.float)[None, :, None, None, None].expand(1, 3, 3, 3, 3),
+            "target": torch.arange(0, 3, dtype=torch.float)[None, :, None, None, None].expand(1, 3, 3, 3, 3) ** 2,
+        },
+        -0.50272596,
+    ],
+]
+
+
+class TestBendingEnergy(unittest.TestCase):
```
Comment (wyli) [inline tests/test_local_normalized_cross_correlation_loss.py:93]: class name out of date

File: monai/losses/image_dissimilarity.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        f2 = (
+            torch.ones([self.in_channels, 1] + [fsize] * self.ndim, dtype=torch.float) / fsize
+        )  # (1, in_channels, D, H, W)
+        # (1, 1, D, H, W) -> (1, in_channels, D, H, W)
+        fn = conv_dict[self.ndim]
+        kernel = fn(f1, f2, padding=int((fsize - 1) / 2))
+
+        return kernel, torch.sum(kernel ** 2), int((fsize - 1) / 2)
+
+    def make_gaussian_kernel(self):
```
Comment (wyli) [inline monai/losses/image_dissimilarity.py:79]: just wanted to mention this in case it's useful: https://github.com/Project-MONAI/MONAI/blob/db7bf8984712f3e46ed533611a70c9b8cd0293bd/monai/networks/layers/convutils.py#L81-L85

File: monai/losses/image_dissimilarity.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            input.ndim - 2 == self.ndim
+        ), f"expecting input with {self.ndim} spatial dimensions, got input of shape {input.shape}"
+        assert (
+            target.shape == input.shape
+        ), f"ground truth has differing shape ({target.shape}) from input ({input.shape})"
+
+        t2, p2, tp = target ** 2, input ** 2, target * input
+
+        # sum over kernel
+        fn = conv_dict[self.ndim]
```
Comment (wyli) [inline monai/losses/image_dissimilarity.py:122]: maybe this could be done in the class constructor (`self.fn = conv_dict[self.ndim]`) to avoid `conv_dict` lookups in every forward

File: monai/losses/image_dissimilarity.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
+import torch
```
Comment (wyli) [inline monai/losses/image_dissimilarity.py:1]: please add license information https://github.com/Project-MONAI/MONAI/blob/db7bf8984712f3e46ed533611a70c9b8cd0293bd/monai/losses/deform.py#L1-L11

File: monai/losses/image_dissimilarity.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        # assume we sum over N values
+        # num = sum[t * p - mean[t] * p - t * mean[p] + mean[t] * mean[p]]
+        #     = sum[t*p] - sum[t] * sum[p] / N * 2 + sum[t] * sum[p] / N
+        #     = sum[t*p] - sum[t] * sum[p] / N
+        #     = sum[t*p] - sum[t] * mean[p] = cross
+        # the following is actually squared ncc
+        cross = tp_sum - p_avg * t_sum
+        t_var = t2_sum - t_avg * t_sum  # std[t] ** 2
+        p_var = p2_sum - p_avg * p_sum  # std[p] ** 2
+        ncc = (cross * cross + EPS) / (t_var * p_var + EPS)  # shape = (batch, 1, D, H, W)
```
Comment (wyli) [inline monai/losses/image_dissimilarity.py:145]: perhaps make this EPS tunable, `self.eps` and defaulting to `1e-7`, e.g. https://github.com/Project-MONAI/MONAI/blob/db7bf8984712f3e46ed533611a70c9b8cd0293bd/monai/losses/dice.py#L51-L52

File: monai/losses/image_dissimilarity.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        ), f"expecting input with {self.ndim} spatial dimensions, got input of shape {input.shape}"
+        assert (
+            target.shape == input.shape
+        ), f"ground truth has differing shape ({target.shape}) from input ({input.shape})"
+
+        t2, p2, tp = target ** 2, input ** 2, target * input
+
+        # sum over kernel
+        fn = conv_dict[self.ndim]
+        t_sum = fn(target, weight=self.kernel, padding=self.padding)
```
Comment (wyli) [inline monai/losses/image_dissimilarity.py:123]: perhaps separable filtering is faster in some cases, would be interesting to have a benchmarking? https://github.com/Project-MONAI/MONAI/blob/db7bf8984712f3e46ed533611a70c9b8cd0293bd/monai/networks/layers/simplelayers.py#L166-L168

**Review by wyli — COMMENTED** [review]
many thanks! I put some minor comments inline

File: monai/losses/image_dissimilarity.py (hunk 2 above)
Comment (kate-sann5100) [reply monai/losses/image_dissimilarity.py:79]: Thanks for mentioning!
Actually, the DeepReg implementation of gaussian kernel uses `(self.kernel_size - 1) / 2.0` as the channel-wise mean out of unknown reason. 
Do we want to correct this by using MONAI's `gaussian_1d` or stick with DeepReg's implementation so we can compare results with DeepReg for unit test?

File: monai/losses/image_dissimilarity.py (hunk 6 above)
Comment (kate-sann5100) [reply monai/losses/image_dissimilarity.py:123]: I will try that : )

File: tests/test_local_normalized_cross_correlation_loss.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
+import unittest
```
[the remark this answers is not in the document]
Comment (kate-sann5100) [reply tests/test_local_normalized_cross_correlation_loss.py:12]: added

File: tests/test_local_normalized_cross_correlation_loss.py (hunk 1 above)
Comment (kate-sann5100) [reply tests/test_local_normalized_cross_correlation_loss.py]: updated

File: monai/losses/image_dissimilarity.py (hunk 3 above)
Comment (kate-sann5100) [reply monai/losses/image_dissimilarity.py]: updated

File: monai/losses/image_dissimilarity.py (hunk 4 above)
Comment (kate-sann5100) [reply monai/losses/image_dissimilarity.py:12]: added

File: monai/losses/image_dissimilarity.py (hunk 5 above)
Comment (kate-sann5100) [reply monai/losses/image_dissimilarity.py]: updated

File: monai/losses/image_dissimilarity.py (hunk 2 above)
Comment (wyli) [reply monai/losses/image_dissimilarity.py:94]: thanks, perhaps we keep the deepreg one for now since we know it works. we could always improve the impl. with another follow-up PR

Comment (YipengHu) [reply monai/losses/image_dissimilarity.py:94]: > Thanks for mentioning!
> Actually, the DeepReg implementation of gaussian kernel uses `(self.kernel_size - 1) / 2.0` as the channel-wise mean out of unknown reason.
> Do we want to correct this by using MONAI's `gaussian_1d` or stick with DeepReg's implementation so we can compare results with DeepReg for unit test?
> @YipengHu

They uses different origins to sample probability densities, should be equivalent (if implemented correctly).

File: monai/losses/image_dissimilarity.py (hunk 6 above)
Comment (YipengHu) [reply monai/losses/image_dissimilarity.py]: > perhaps separable filtering is faster in some cases, would be interesting to have a benchmarking?
> 
> https://github.com/Project-MONAI/MONAI/blob/db7bf8984712f3e46ed533611a70c9b8cd0293bd/monai/networks/layers/simplelayers.py#L166-L168

yup - pity that deepreg had separable gaussian kernel implemented in a file next to it ;) @acasamitjana @mathpluscode, a ticket?

Comment (mathpluscode) [reply monai/losses/image_dissimilarity.py]: actually, we still have it https://github.com/DeepRegNet/DeepReg/blob/7c22a30ef27d7ff1bec5d21cbdb1a25f65cd5027/deepreg/model/loss/label.py#L314

Comment (acasamitjana) [reply monai/losses/image_dissimilarity.py]: I think separable convolutions for the three kernels can be interesting. I have just opened a ticket on DeepReg and we can port it here

File: monai/losses/image_dissimilarity.py (hunk 2 above)
Comment (kate-sann5100) [reply monai/losses/image_dissimilarity.py:94]: > > Thanks for mentioning!
> > Actually, the DeepReg implementation of gaussian kernel uses `(self.kernel_size - 1) / 2.0` as the channel-wise mean out of unknown reason.
> > Do we want to correct this by using MONAI's `gaussian_1d` or stick with DeepReg's implementation so we can compare results with DeepReg for unit test?
> > @YipengHu
> 
> They use different origins to sample probability densities, should be equivalent (if implemented correctly).

The monai `gaussian_1d` enforces centre distribution (the centre pixel of the kernel to be 0), while in DeepReg, the kernel is not symmetric along the channel dimension.
Do you think it will actually be better to enforce centre distribution on the channel dimension?

File: monai/losses/image_dissimilarity.py (hunk 6 above)
Comment (kate-sann5100) [reply monai/losses/image_dissimilarity.py]: separable filter implemented

**Review by wyli — APPROVED** [review]
thanks!

File: monai/losses/image_dissimilarity.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        else:
+            raise ValueError(
+                f'Unsupported kernel_type: {kernel_type}, available options are ["rectangular", "triangular", "gaussian"].'
+            )
+
+        self.kernel_vol = torch.sum(self.kernel) ** self.ndim
+        self.smooth_nr = float(smooth_nr)
+        self.smooth_dr = float(smooth_dr)
+
+    def make_rectangular_kernel(self):
```
Comment (mathpluscode) [inline monai/losses/image_dissimilarity.py:85]: These kernel functions requires only the argument `kernel_size`.

I suggest to move these functions outside of the class, or mark them as `@staticmethod`.

File: monai/losses/image_dissimilarity.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+        self.ndim = ndim
+        if self.ndim not in [1, 2, 3]:
+            raise ValueError(f"Unsupported ndim: {self.ndim}-d, only 1-d, 2-d, and 3-d inputs are supported")
+
+        self.kernel_size = kernel_size
+        if self.kernel_size % 2 == 0:
+            raise ValueError(f"kernel_size must be odd, got {self.kernel_size}")
+
+        if kernel_type == "rectangular":
```
Comment (mathpluscode) [inline monai/losses/image_dissimilarity.py:70]: I suggest to keep a dictionary as a class attribute and use dict instead of if/else. similar to https://github.com/DeepRegNet/DeepReg/blob/main/deepreg/model/loss/image.py#L241.

Or maybe MONAI has different requirements ^^

File: monai/losses/image_dissimilarity.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+            self.kernel = self.make_rectangular_kernel()
+        elif kernel_type == "triangular":
+            self.kernel = self.make_triangular_kernel()
+        elif kernel_type == "gaussian":
+            self.kernel = self.make_gaussian_kernel()
+        else:
+            raise ValueError(
```
Comment (mathpluscode) [inline monai/losses/image_dissimilarity.py:77]: this error is not tested

File: monai/losses/image_dissimilarity.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+            kernel_size: kernel spatial size, must be odd.
+            kernel_type: {``"rectangular"``, ``"triangular"``, ``"gaussian"``}. Defaults to ``"rectangular"``.
+            reduction: {``"none"``, ``"mean"``, ``"sum"``}
+                Specifies the reduction to apply to the output. Defaults to ``"mean"``.
+
+                - ``"none"``: no reduction will be applied.
+                - ``"mean"``: the sum of the output will be divided by the number of elements in the output.
+                - ``"sum"``: the output will be summed.
+            smooth_nr: a small constant added to the numerator to avoid zero.
+            smooth_dr: a small constant added to the denominator to avoid nan.
```
Comment (mathpluscode) [inline monai/losses/image_dissimilarity.py:57]: Do we need to mention raised error in docstring?

Comment (mathpluscode) [reply monai/losses/image_dissimilarity.py:57]: the format would be https://google.github.io/styleguide/pyguide.html#244-decision

**Review by mathpluscode — CHANGES_REQUESTED** [review]
Suggested some refactoring and raised errors should be covered in the tests.
But not sure if we need to mention potential errors in docstring, it'd be better to do so.

File: monai/losses/image_dissimilarity.py (hunk 11 above)
Comment (kate-sann5100) [reply monai/losses/image_dissimilarity.py:57]: This complies with the implementation of other losses.

File: monai/losses/image_dissimilarity.py (hunk 10 above)
Comment (kate-sann5100) [reply monai/losses/image_dissimilarity.py:99]: Done

File: monai/losses/image_dissimilarity.py (hunk 9 above)
Comment (kate-sann5100) [reply monai/losses/image_dissimilarity.py]: Done

File: monai/losses/image_dissimilarity.py (hunk 8 above)
Comment (kate-sann5100) [reply monai/losses/image_dissimilarity.py]: Done

**Review by mathpluscode — APPROVED** [review]

File: monai/losses/image_dissimilarity.py (hunk 9 above)
Comment (wyli) [reply monai/losses/image_dissimilarity.py]: monai extends `Enum` https://github.com/Project-MONAI/MONAI/blob/a4ef69166c88ab252ef06268bc4c8fc7a089e36b/monai/utils/enums.py#L34 but this style is not yet enforced
